sdk/secp256r1/src/lib.rs

```diff
@@ -0,0 +1,674 @@
+
+        // We do not check or use the byte at data[1]
+        if data.len() < expected_data_size {
+            return Err(PrecompileError::InvalidInstructionDataSize);
+        }
+
+        // Parse half order from constant
+        let half_order: BigNum = BigNum::from_slice(&SECP256R1_HALF_ORDER)
+            .map_err(|_| PrecompileError::InvalidSignature)?;
+
+        // Parse order - 1 from constant
+        let order_minus_one: BigNum = BigNum::from_slice(&SECP256R1_ORDER_MINUS_ONE)
+            .map_err(|_| PrecompileError::InvalidSignature)?;
+
+        // Create a BigNum for 1
+        let one = BigNum::from_u32(1).map_err(|_| PrecompileError::InvalidSignature)?;
+
+        // Define curve group
+        let group = EcGroup::from_curve_name(Nid::X9_62_PRIME256V1)
+            .map_err(|_| PrecompileError::InvalidSignature)?;
+        let mut ctx = BigNumContext::new().map_err(|_| PrecompileError::InvalidSignature)?;
+
+        for i in 0..num_signatures {
+            let start = i
+                .saturating_mul(SIGNATURE_OFFSETS_SERIALIZED_SIZE)
+                .saturating_add(SIGNATURE_OFFSETS_START);
+            let end = start.saturating_add(SIGNATURE_OFFSETS_SERIALIZED_SIZE);
+
+            // bytemuck wants structures aligned
+            let offsets: &Secp256r1SignatureOffsets =
+                bytemuck::try_from_bytes(&data[start..end])
+                    .map_err(|_| PrecompileError::InvalidDataOffsets)?;
+
+            // Parse out signature
+            let signature = get_data_slice(
+                data,
+                instruction_datas,
+                offsets.signature_instruction_index,
+                offsets.signature_offset,
+                SIGNATURE_SERIALIZED_SIZE,
+            )?;
+
+            // Parse out pubkey
+            let pubkey = get_data_slice(
+                data,
+                instruction_datas,
+                offsets.public_key_instruction_index,
+                offsets.public_key_offset,
+                COMPRESSED_PUBKEY_SERIALIZED_SIZE,
+            )?;
+
+            // Parse out message
+            let message = get_data_slice(
+                data,
+                instruction_datas,
+                offsets.message_instruction_index,
+                offsets.message_data_offset,
+                offsets.message_data_size as usize,
+            )?;
+
+            let r_bignum = BigNum::from_slice(&signature[..FIELD_SIZE])
+                .map_err(|_| PrecompileError::InvalidSignature)?;
+            let s_bignum = BigNum::from_slice(&signature[FIELD_SIZE..])
+                .map_err(|_| PrecompileError::InvalidSignature)?;
+
+            // Check that the signature is generally in range
+            let within_range = r_bignum >= one
+                && r_bignum <= order_minus_one
+                && s_bignum >= one
+                && s_bignum <= half_order;
+
+            if !within_range {
+                return Err(PrecompileError::InvalidSignatureRange);
```

Comment:
I think it is not ideal to have both a generic and specific error variants for invalid signatures. Let's just keep it simple and use `PrecompileError::InvalidSignature` here 🙏